src/Workspaces/Core/Portable/Shared/Utilities/RoslynParallel.NetFramework.cs

```diff
@@ -0,0 +1,655 @@
+                        // is responsible only for creating the next worker, which in turn means there can't be any contention
+                        // on creating workers (though it's possible one worker could be executing while we're creating the next).
+                        if (!launchedNext)
+                        {
+                            launchedNext = true;
+                            state.QueueWorkerIfDopAvailable();
+                        }
+
+                        // Process the loop body.
+                        await state.LoopBody(element, state.Cancellation.Token);
+                    }
+                }
+                catch (Exception e)
+                {
+                    // Record the failure and then don't let the exception propagate.  The last worker to complete
+                    // will propagate exceptions as is appropriate to the top-level task.
+                    state.RecordException(e);
+                }
+                finally
+                {
+                    // If we're the last worker to complete, clean up and complete the operation.
+                    if (state.SignalWorkerCompletedIterating())
+                    {
+                        try
+                        {
+                            await state.DisposeAsync();
+                        }
+                        catch (Exception e)
+                        {
+                            state.RecordException(e);
+                        }
+
+                        // Finally, complete the task returned to the ForEachAsync caller.
+                        // This must be the very last thing done.
+                        state.Complete();
+                    }
+                }
+#pragma warning restore CA2007
+            };
+
+            try
+            {
+                // Construct a state object that encapsulates all state to be passed and shared between
+                // the workers, and queues the first worker.
+                var state = new AsyncForEachAsyncState<TSource>(source, taskBody, dop, scheduler, cancellationToken, body);
+                state.QueueWorkerIfDopAvailable();
+                return state.Task;
+            }
+            catch (Exception e)
+            {
+                return Task.FromException(e);
+            }
+        }
+
+        /// <summary>Gets the default degree of parallelism to use when none is explicitly provided.</summary>
+        private static int DefaultDegreeOfParallelism => Environment.ProcessorCount;
+
+        /// <summary>Stores the state associated with a ForEachAsync operation, shared between all its workers.</summary>
+        /// <typeparam name="TSource">Specifies the type of data being enumerated.</typeparam>
+#if false
+        private abstract class ForEachAsyncState<TSource> : TaskCompletionSource, IThreadPoolWorkItem
+#else
+        private abstract class ForEachAsyncState<TSource> : TaskCompletionSource<VoidResult>
```

Comment:
i think it should be safe to drop IThreadPoolWorkItem.  Though i wouldn't mind @stephentoub  confirming.

Switched to the generic TCS as that's all that netfx has.

Reply:
> i think it should be safe to drop IThreadPoolWorkItem. Though i wouldn't mind @stephentoub confirming.

On .NET Core that serves as an optimization to avoid allocating each time it's queued. It's functionally fine to drop it.

Reply:
sgtm.